pkg/store/bucket.go

```diff
@@ -295,59 +294,38 @@ func (s *BucketStore) Close() (err error) {
 // SyncBlocks synchronizes the stores state with the Bucket bucket.
 // It will reuse disk space as persistent cache based on s.dir param.
 func (s *BucketStore) SyncBlocks(ctx context.Context) error {
-	var wg sync.WaitGroup
-	blockc := make(chan ulid.ULID)
-
-	for i := 0; i < s.blockSyncConcurrency; i++ {
-		wg.Add(1)
-		go func() {
-			for id := range blockc {
-				if err := s.addBlock(ctx, id); err != nil {
-					level.Warn(s.logger).Log("msg", "loading block failed", "id", id, "err", err)
-					continue
-				}
-			}
-			wg.Done()
-		}()
-	}
-
 	allIDs := map[ulid.ULID]struct{}{}
 
-	err := s.bucket.Iter(ctx, "", func(name string) error {
-		// Strip trailing slash indicating a directory.
-		id, err := ulid.Parse(name[:len(name)-1])
-		if err != nil {
-			return nil
+	mds, err := s.blockMDFetcher.Fetch(ctx)
+	if err != nil {
+		return errors.Wrap(err, "fetching remote block metadata failed")
+	}
+	for blockID, md := range mds {
+		if md == nil {
+			continue
 		}
 
-		inRange, err := s.isBlockInMinMaxRange(ctx, id)
+		inRange, err := s.isBlockInMinMaxRange(ctx, blockID)
 		if err != nil {
-			level.Warn(s.logger).Log("msg", "error parsing block range", "block", id, "err", err)
-			return nil
+			level.Warn(s.logger).Log("msg", "error parsing block range", "block", blockID, "err", err)
+			continue
 		}
 
 		if !inRange {
-			return nil
+			continue
 		}
 
-		allIDs[id] = struct{}{}
+		allIDs[blockID] = struct{}{}
 
-		if b := s.getBlock(id); b != nil {
-			return nil
-		}
-		select {
-		case <-ctx.Done():
-		case blockc <- id:
+		if b := s.getBlock(blockID); b != nil {
+			continue
 		}
-		return nil
-	})
-
-	close(blockc)
-	wg.Wait()
 
-	if err != nil {
-		return errors.Wrap(err, "iter")
+		if err := s.addBlock(ctx, blockID); err != nil {
```

Comment:
This is important as we download meta.json twice. Once to memory once to disk... Can we improve this?